### PR TITLE
minitron pruning

### PR DESCRIPTION
**Pull request by m1kush**

### CODE REVIEW

File: src/core/checkpointing.py [hunk 1]
```diff
@@ -12,16 +12,18 @@
...
         self.model = model
         self.optimizer = optimizer
         self.scheduler = scheduler
+        self.no_optimizer = no_optimizer
 
     def state_dict(self):
         # this line automatically manages FSDP FQN's, as well as sets the default state dict type to FSDP.SHARDED_STATE_DICT
         model_state_dict, optimizer_state_dict = get_state_dict(
-            self.model, self.optimizer
+            self.model, self.optimizer if not self.no_optimizer else []
```
Comment (crewtool) [inline src/core/checkpointing.py:24]: self.optimizer if self.optimizer else []

File: src/core/checkpointing.py [hunk 2]
```diff
@@ -130,7 +133,8 @@ def load_checkpoint_from_file(load_config, model, optimizer, scheduler):
     if checkpoint_path is not None:
         if isinstance(model, FSDP):
             # Sharded load
-            state_dict = {"app": TrainingState(model, optimizer, scheduler)}
+            no_optimizer = load_config.get("load_without_optimizer", False)
```
Comment (crewtool) [inline src/core/checkpointing.py:136]: This would not be needed

File: src/projected_compression/minitron.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import torch.nn as nn
+import json
+import logging
+import os
+
+from src.core.checkpointing import get_full_checkpoint_path
+
+logger = logging.getLogger(__name__)
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (crewtool) [inline src/projected_compression/minitron.py:11]: We should have one place for setting device

File: src/projected_compression/minitron.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    logger.debug("Importance dimensions calculated.")
+
+    dmodel_top_indices = torch.topk(dmodel_importance, dim=0, largest=True, k=target_dmodel).indices.tolist()
+
+    dff_top_indices = []
+    for i in range(n_blocks):
+        dff_top_indices_current = torch.topk(dff_importance[i], dim=0, largest=True, k=target_dff).indices.tolist()
+        dff_top_indices.append(dff_top_indices_current)
+
+    # save to file indices as dict
```
Comment (crewtool) [inline src/projected_compression/minitron.py:102]: Is this comment chatgpt genersted?

File: configs/fine_tune_llama_hf.yaml
Comment (crewtool) [inline configs/fine_tune_llama_hf.yaml:1]: Remove

File: configs/_trainer/default.yaml
Comment (crewtool) [inline configs/_trainer/default.yaml:1]: Remove

File: src/projected_compression/minitron.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    dict_to_save = {"dmodel_top_indices": dmodel_top_indices, "dff_top_indices": dff_top_indices}
+    path = get_full_checkpoint_path(checkpoint_save_path) + "/top_indices.json"
+
+    # check if path exists
```
Comment (crewtool) [inline src/projected_compression/minitron.py:106]: Remove

File: src/projected_compression/minitron.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    os.makedirs(os.path.dirname(path), exist_ok=True)
+
+    with open(path, "w") as f:
+        json.dump(dict_to_save, f)
```
Comment (crewtool) [inline src/projected_compression/minitron.py:110]: How about torch.save?

File: configs/_checkpoints/mas/hf_llama_3_1.yaml
Comment (madragonse) [inline configs/_checkpoints/mas/hf_llama_3_1.yaml:1]: can we incorporate new configs from from #47 ?

File: grid_generator/sbatch_builder.py
Comment (madragonse) [inline grid_generator/sbatch_builder.py:1]: comparison to main error? - changes from other PRs looks like

File: src/core/checkpointing.py [hunk 7]
```diff
@@ -12,16 +12,18 @@
 logger = logging.getLogger(__name__)
 
 class TrainingState(Stateful):
-    def __init__(self, model, optimizer, scheduler):
+    def __init__(self, model, optimizer, scheduler, no_optimizer=False):
```
[the remark this answers is not in the document]
Comment (madragonse) [reply src/core/checkpointing.py:15]: can be usefull in the future, but at the moment dont need, can make problems in loading, sometimes optimizer/scheduler dont have any info, but better saved than sorry,
